src/Sparrow/AsyncManualResetEvent.cs

```diff
@@ -91,13 +91,23 @@ public async Task<bool> WaitAsync(TimeSpan timeout)
                 var waitAsync = _tcs.Task;
                 _parent._token.ThrowIfCancellationRequested();
                 var result = await Task.WhenAny(waitAsync, TimeoutManager.WaitFor(timeout, _parent._token)).ConfigureAwait(false);
+
+                if (result.IsFaulted)
+                    throw result.Exception;
+
                 if (_parent._token != CancellationToken.None)
                     return result == waitAsync && !_parent._token.IsCancellationRequested;
 
                 return result == waitAsync;
             }
         }
 
+        public void SetException(Exception e)
+        {
+            _token.ThrowIfCancellationRequested();
+            _tcs.SetException(e);
```

Comment:
We should use here `TrySetException`

Reply:
Maybe we should add another method for that (TrySetException)